libs/utils/src/lock_file.rs

```diff
@@ -0,0 +1,82 @@
+//! A module to allow processes to lock on specific FS file exclusive.
+//! Could be used to forbid multiple processes to start in the same directory.
+//!
+//! Note that locked files are done via [`fcntl::flock`] which means that
+//! * files are locked only for taking another lock on them: all other operations are still possible on files.
+//! For example, other process could do `rm file` and that would be allowed.
+//! * if no file exists (after `rm` in another process), the process that "locked" the file still runs and does not error
+//! new procesess are able to start in the same directory, as they will be creating another file to lock.
+
+use std::{
+    fs,
+    os::unix::prelude::{AsRawFd, OpenOptionsExt},
+    path::Path,
+};
+
+use anyhow::Context;
+use nix::{fcntl, libc};
+
+use crate::crashsafe;
+
+pub enum LockCreationResult {
+    Created {
+        new_lock_contents: String,
+        file: fs::File,
+    },
+    AlreadyLocked {
+        existing_lock_contents: String,
+    },
+    CreationFailed(anyhow::Error),
+}
+
+/// Creates a lock file in the path given and writes the given contents into the file.
+/// Note: The lock is automatically released when the file closed. You might want to use Box::leak to make sure it lives until the end of the program.
+pub fn create_lock_file(lock_file_path: &Path, contents: String) -> LockCreationResult {
```

Comment:
Come to think of it, let's just Box::leak() it right here. Why leave it to the callers, if that's what we expect all the callers to do anyway? If we need a lock file that doesn't live until the end of the process in the future, we can refactor it then.

Reply:
I would like to leave this as is, since we're supposed to create .pid file for etcd and potentially other processes.
We don't want to leak that locked file there, only create one safely, checking for locks and that's why I reuse `create_lock_file` for that and do not leak the final file.